### PR TITLE
Ideogram 4.0

### PR DESCRIPTION
**Pull request by cglagovichTT**
### PR Category
Feature

Add support for Ideogram 4.0 image generation model.

E2E performance — 2048px V4_TURBO_12 (program-cache warm, SP4×TP2 LoudBox)
| Module | Time (s) |
| --- | --- |
| encoder | 0.37 |
| denoising | 17.84 |
| vae | 4.05 |
| total | 23.64 |

Example images:
ideogram4_2048_wild_typography
ideogram4_2048_mango_sago

### Summary
Brings up **Ideogram 4.0** text-to-image on Blackhole in `tt_dit`, composing existing ttnn / tt_dit ops (no new kernels). The pipeline has four components:

- **Text encoder** — a new Qwen3-VL-8B text encoder (`encoders/qwen3vl`) with a 13-layer feature tap; per-head QK-norm and MRoPE, TP + optional FSDP on the non-TP axis.
- **Denoiser** — a single-stream flow-matching DiT (`models/transformers/transformer_ideogram4.py`, 34 blocks, `emb_dim` 4608, 18 heads @ head_dim 256, SwiGLU, 4-branch tanh-gated AdaLN, per-head QK-RMSNorm, interleaved MRoPE). Wan-style TP-fractured residual + ring sequence parallelism.
- **VAE decoder** — reuses the SD3.5 `VAEDecoder`, adding only `post_quant_conv` (`models/vae/vae_ideogram4.py`).
- **Sampler** — logit-normal Euler flow-matching schedule with per-step guidance presets `V4_TURBO_12` / `V4_DEFAULT_20` / `V4_QUALITY_48` (`pipelines/ideogram4/sampler.py`).

**Asymmetric CFG:** the conditional and unconditional passes are distinct distilled nets, so (unlike the cfg-parallel Flux/Qwen-Image pipelines) CFG is not split across submeshes — both nets stay resident on the full mesh and run sequentially, with the velocity blended on-device (`ttnn.lerp`). The full per-step computation (both forwards + velocity gathers + CFG blend + Euler update) runs in a single trace with the latent kept device-resident across the loop.

Runs on the Blackhole LoudBox (4×2, SP4×TP2, Linear) and Galaxy (4×8, TP4×SP8, Ring), discovered from the mesh shape via pipeline presets. Weights load from the gated Ideogram 4.0 fp8 checkpoint (`IDEOGRAM4_WEIGHTS`), dequantized to bf16 with an on-disk cache.

Shared-infra changes (all additive / backwards-compatible):
- `layers/normalization.py`: `DistributedRMSNorm` gains `per_head_norm` (independent RMS over each head's head_dim) and `dynamic_weight` (folds an adaLN `(1 + scale)` into the fused op's affine). Default behavior for existing callers is unchanged.
- `layers/conv2d.py`: fall back to a documented max DRAM width-slice count for conv shapes not in the tuned table, and spill conv config tensors to DRAM for very large convs (VAE decode at 2048px). Existing tuned shapes unaffected.
- `dit_fused_distributed_rmsnorm_program_factory.cpp`: L1-capacity accounting for the pre-existing `per_head_norm` path; gated behind the flag, existing callers unchanged.

### Notes for reviewers
- **Reference port:** `reference/ideogram4/modeling_ideogram4.py` is the PyTorch golden (adapted from the Ideogram OSS release, Apache-2.0). The tt_dit block mirrors it op-for-op; worth a pass to confirm the AdaLN / SwiGLU / masking-before-projection semantics match.
- **LLM feature-tap ordering:** `interleave_layer_taps` assembles the 13 Qwen3-VL layer taps feature-major (`out[..., f*13 + l]`) — this is the layout the trained `llm_cond_proj` expects and is locked by a host unit test. Please sanity-check this is the intended order.
- **`per_head_norm` / `dynamic_weight`** on the shared `DistributedRMSNorm` are the main cross-model touch points — confirm the fused-op semantics (batched per-sample weight broadcast, per-head local reduction) are safe for the other callers (WAN2.2 / LTX).
- **Test coverage:** per-component PCC tests exist for the encoder (per-layer), block, full transformer, VAE, sampler, and CFG blend, with SP/TP sweeps on the block and VAE. There is intentionally **no end-to-end pipeline PCC test** (the whole-pipeline test is a shape/std smoke test), and real-weight tests skip unless `IDEOGRAM4_WEIGHTS` is set — flagging in case reviewers want a gated end-to-end golden.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:cglagovich/ideogram4-mainline)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:cglagovich/ideogram4-mainline)

### CODE REVIEW

File: models/tt_dit/tests/models/ideogram4/test_pipeline_class_ideogram4.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+# the denoiser matmuls use none, so 64KB over-reserved the shared L1 region and made the SP4xTP2
+# path's resident ring/all-gather buffers clash with the matmul circular buffers.
+_LINE_PIPE = {**line_params, "l1_small_size": 32768, "trace_region_size": 60000000}
+_RING_PIPE = {**ring_params, "l1_small_size": 32768, "trace_region_size": 60000000}
+
+
+# Pipeline mesh configs: the parallel config (tp/sp/num_links/topology) is DISCOVERED from the
+# mesh shape via Ideogram4Pipeline._PRESETS, so tests only pick a mesh shape + its device_params.
+_PIPE_CONFIGS = [
+    pytest.param((4, 2), _LINE_PIPE, id="sp4tp2"),  # BH loudbox: SP4 x TP2
```
Comment (kevinmi-TT) [inline models/tt_dit/tests/models/ideogram4/test_pipeline_class_ideogram4.py]: should we keep the 2x4, 4x8 convention?

File: models/tt_dit/encoders/qwen3vl/__init__.py
Comment (kevinmi-TT) [inline models/tt_dit/encoders/qwen3vl/__init__.py:1]: did we needed this init

File: models/tt_dit/pipelines/ideogram4/pipeline_ideogram4.py [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+        self.topology = config.topology
+        weights_dir = config.weights_dir
+        if not weights_dir:
+            raise ValueError(
+                "Ideogram4 weights directory not set: pass weights_dir=... to create_pipeline "
+                "or set the IDEOGRAM4_WEIGHTS environment variable to the fp8 checkpoint dir."
+            )
+        self.weights_dir = weights_dir
+        qwen_repo = config.qwen_repo
+        self.config = modeling_ideogram4.Ideogram4Config()
```
Comment (kevinmi-TT) [inline models/tt_dit/pipelines/ideogram4/pipeline_ideogram4.py]: why is there two config and _config, a bit confusing

File: models/tt_dit/encoders/qwen3vl/model_qwen3vl.py [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+            math_fidelity=ttnn.MathFidelity.HiFi4,
+            math_approx_mode=False,
+            fp32_dest_acc_en=True,
+        )
+
+    def forward(self, x: ttnn.Tensor) -> ttnn.Tensor:
+        return super().forward(x, compute_kernel_config=self._compute_kernel_config)
+
+
+def _apply_rope(x: ttnn.Tensor, cos: ttnn.Tensor, sin: ttnn.Tensor) -> ttnn.Tensor:
```
Comment (kevinmi-TT) [inline models/tt_dit/encoders/qwen3vl/model_qwen3vl.py:501]: could just apply the same split rope convert to interleave like the denosier

**Review by kevinmi-TT — APPROVED** [review]

File: models/tt_dit/tests/models/ideogram4/test_pipeline_class_ideogram4.py (hunk 1 above)
Comment (cglagovichTT) [reply models/tt_dit/tests/models/ideogram4/test_pipeline_class_ideogram4.py]: fixed, thanks

**Review by jonathansuTT — APPROVED** [review]
best model ever

**Review by llongTT — APPROVED** [review]
